File: python/09_queue/array_queue.py
"""
    Queue based upon array
    用数组实现的队列

    Author: Wenru
"""

from typing import Optional

class ArrayQueue:
    
    def __init__(self, capacity: int):
        self._items = []
        self._capacity = capacity
        self._head = 0
        self._tail = 0

    def enqueue(self, item: str) -> bool:
        if self._tail == self._capacity:
            if self._head == 0:
                return False
            else:
                for i in range(0, self._tail - self._head):
                    self._items[i] = self._items[i + self._head]
                self._tail = self._tail - self._head
                self._head = 0
        
        self._items.insert(self._tail, item)
        self._tail += 1
        return True
    
    def dequeue(self) -> Optional[str]:
        if self._head != self._tail:
            item = self._items[self._head]
            self._head += 1
            return item
        else:
            return None
    
    def __repr__(self) -> str:
        return " ".join(item for item in self._items[self._head : self._tail])

File: kotlin/09_queue/ArrayQueue.kt
package com.wil.leetcode.algo_09_queue

/**
 *
 * Desc:用数组实现的队列,申请容量大小的数组
 * <p>
 * Date: 2019-12-05
 * Updater:
 * Update Time:
 * Update Comments:
 *
 * Author: pengyushan
 */
class ArrayQueue(private val capacity: Int) {
    // 定义容量大小的数组
    private val items: Array<String?> = arrayOfNulls(capacity)
    // 队头下标
    private var head = 0
    //队尾下标
    private var tail = 0

    // 入队
    fun enqueue(item: String?): Boolean {
        // 如果队尾已经等于n了，表示队列已经满了
        if (tail == capacity) return false
        items[tail] = item
        ++tail
        return true
    }

    // 出队
    fun dequeue(): String? {
        // 如果head == tail 表示队列为空
        if (head == tail) return null
        val ret = items[head]
        ++head
        return ret
    }
}